### DISCUSSION TITLE
About App Store Guidelines

### DISCUSSION THREAD

**Discussion by gdelataillade** [gdelataillade/alarm#87] [General]
For those who tried to publish their app containing the plugin `alarm` to the App Store, I'm curious, did you face any difficulties ?

**Comment by synstin** [comment]
They asked what part of the audio was playing in the background, and when I explained where the audio was playing, they passed.

**Reply by gdelataillade** [reply]
Awesome ! Then your app is available in the App Store ?

**Reply by synstin** [reply]
@gdelataillade 
Yes my app is available for download on the app store.
I was struggling with background processing and this package made it easy and seamless to implement.
This is the best package I've ever used.
Thanks a lot.

**Reply by gdelataillade** [reply]
@synstin Thanks a lot for the kind words! It's awesome to hear my package helped you!

By the way, I'd be happy to feature your app in my package's README. It could give your app some extra spotlight and show others a cool example of the package in action. Just let me know if you're up for it!

**Reply by alaslipknot** [reply]
@synstin any chance you can share a link for your app ? am curious to see this package in action.

**Reply by gdelataillade** [reply]
I just created a discussion where developers can share their apps: #170 

@synstin feel free to post in the discussion if you want
@alaslipknot check out the discussion in a few days to see if apps were shared

**Comment by fimpact-nicole** [comment]
Initially, my app was rejected due to questions regarding the necessity of the `UIBackgroundModes` key (Guideline 2.5.4 - Performance - Software Requirements). After explaining the necessity of `UIBackgroundModes` with reference to the comments in  #96, that problem was resolved.

Then, my app was rejected again with another reason: `"Your app plays an alarm but does not display the notification to access your app as outlined in your previous response (Review device: iPhone 13 mini & iOS 17.4.1)"` (Guideline 2.1 - Performance).

I tested on various physical devices and simulators, but the issue was not reproduced. Therefore, I have replied asking the reviewer for more details or steps to reproduce the issue.

**Reply by gdelataillade** [reply]
It's weird that they didn't receive the notification. Do you prompt for notification permission in your app ?

**Reply by fimpact-nicole** [reply]
@gdelataillade 
I couldn't reproduce the issue with simulators and physical devices, so I explained again more precisely about the difference between the case when my app is in the foreground state and in the background state and requested more details to reproduce the issue.

And then, today, the reviewer replied again that the `UIBackgroundModes` key with "audio" setting was incorrect:
```
Regarding Guideline 2.5.4, background audio intended for setting and playing alarms is inappropriate 
because it can be replicated with Local Notifications. 
Background audio is intended for use by apps that provide audible content to the user while in the background, 
such as music player, music creation, or streaming audio apps. 

To resolve this issue, it would be appropriate to remove the "audio" setting from the UIBackgroundModes key, 
if the app does not have a feature that requires persistent audio.
```

I had already explained the necessity of the `UIBackgroundModes` key, referencing the implementation method and using Alarmy as an example.
It is very difficult to pass iOS review. 😂

**Reply by gdelataillade** [reply]
Hi @fimpact-nicole 

Maybe try to tell them that using `audio` is mandatory for your app so you make sure it stays alive for the moment your alarms has to ring. Local notification is not a good option for many reasons: it can't play long audios, it won't show if device is in Do not disturb modes, vibrations are limited too...

Keep trying, I'm sure your app will soon pass the review ! 💪

**Reply by fimpact-nicole** [reply]
@gdelataillade 
Thank you for your response. 
Fortunately, the app passed the review process after that, and version 1.0.0 was released on the [App Store](https://apps.apple.com/us/app/sunrise-alarm-clock-journal/id6502430422)! (There were no issues with the Play Store.)

Version 1.0.0 was an alarm + morning diary app, but regrettably, our team have decided to focus on the diary feature in the next version, which means the alarm functionality might be removed.

However, thanks to the alarm package, I was able to easily implement the alarm feature. Your package allowed me to effortlessly integrate the alarm functionality into my app, saving me a great deal of time and effort. Thank you! 🙇

**Reply by gdelataillade** [reply]
Hi @fimpact-nicole 

I'm glad to hear you liked the plugin and that your app passed the review process !

If you're interested, you can share your app links in this thread #170.

**Comment by vsaase** [comment]
My app was rejected even though I explained the necessity of background mode:

They said:
"We still find that this app uses background audio to keep the app alive in the background and/or for audio alerts.

Background audio is intended for use by apps that provide audible content to the user while in the background, such as music player, music creation, or streaming audio apps."

**Reply by gdelataillade** [reply]
Hi @vsaase 

This is what I told to someone in this discussion and it passed the review:

> Maybe try to tell them that using audio is mandatory for your app so you make sure it stays alive for the moment your alarms has to ring. Local notification is not a good option for many reasons: it can't play long audios, it won't show if device is in Do not disturb modes, vibrations are limited too...
> Keep trying, I'm sure your app will soon pass the review ! 💪

**Comment by NiceGuyNimni** [comment]
Hey,
First of all - amazing work!
Thanks a lot for this superb library!

I'm also having trouble with Apple's approval.
My app functions as a solidarity alarm that works at a specific time of day (17:00) and the user can only toggle on or off the alarm.

**This is the rejection I got:**

> The app declares support for audio in the UIBackgroundModes key in your Info.plist, but we are unable to play any audible content when the app is running in the background.
> Background audio is intended for use by apps that provide audible content to the user while in the background, such as music player, music creation, or streaming audio apps. 

**I've sent them this response:**

> The app is intended to perform as an alarm, more specifically, a solidarity alarm where people can pause for a minute at 17:00 and think of the captives.
> Basically it means playing a sound when the app is not in the foreground since no one will keep the app in the foreground forever and no one will open it manually at 17:00 to hear the alarm sound.
> Thus, using audio in the background is mandatory for my app in order to make sure it stays alive for the moment the alarms has to ring.
> Hopefully, when AlarmKit will be widely supported in most iPhone devices - I'll use it but until then I had to use a different approach in order to be able to play an alarm on a daily basis.

I'll report back once I'll get a response from Apple.

**Reply by gdelataillade** [reply]
Hi @NiceGuyNimni, thanks for the nice post. I hope Apple will approve your app, I think your response is good enough. Keep me updated!

About AlarmKit, I created a package for it for Flutter so it's easy to implement in your app: https://pub.dev/packages/flutter_alarmkit
As you mentioned, iOS 26 is still not widely used so it's not a good solution for now but hopefully it becomes the main solution for alarms soon.
